Rate limit

maxence-charriere opened the pull request:
Hello @rbranson @tsholmes @ndhoule @thehydroimpulse
Here is a first try on the rate implementation for drip.

Right now it is just on the identify call.
Would be glad to have some feedback to see if I’m going on the right
direction.

Also I’m not sure about the re enqueuing mechanics. Any hint would be
welcome.

tsholmes commented on lib/index.js:
```diff
@@ -113,4 +143,70 @@ Drip.prototype.normalize = function(obj){
+  var limit = {
+    total: headers['x-ratelimit-limit'],
+    remain: headers['x-ratelimit-remaining'],
+    reset: Date.now() + (1000 * 60 * 3)
```
They don't include it in the header, but experimentally, the limits reset at hour boundaries. We could set the reset time here a little more accurately (although then we run the risk of it being very wrong if they change their rate limiting implementation)

maxence-charriere commented:
@tsholmes
Just refactored to use setLimit only on OK status and 429.

Being the last condition checked, ```reset``` is set 5 min after from when ```setLimit``` is called . This allows tu regularly check the cooldown has been reset and we would not depend of Radom changes from Drip side.

Let me know if some other thing should be changed.

tsholmes commented on lib/index.js:
```diff
@@ -48,27 +48,46 @@ var Drip = module.exports = integration('Drip')
+          }
+
+          if (res.ok) {
+            return self.setLimit(fn(null, res));
```
This doesn't do what you want I think. This will call the callback function and pass the result to `setLimit`.

tsholmes commented on lib/index.js:
```diff
@@ -48,27 +48,46 @@ var Drip = module.exports = integration('Drip')
+          var msg = errors[0].message;
+          var status = res.status;
+
+          if (res.status == 429) {
```
Should use `===` here and elsewhere for comparisons

tsholmes commented on lib/index.js:
```diff
@@ -48,27 +48,46 @@ var Drip = module.exports = integration('Drip')
+            return fn(null, res);
+          }
+
+          fn(self.error('bad request status=%s msg=%s', status, msg || ''));
```
I think we're probably better off just passing `err` here instead of `self.error(...)`. This will pass the error code out properly and the error message will still be available in the logs

tsholmes commented on test/index.js:
```diff
@@ -90,6 +107,21 @@ describe('Drip', function(){
         .error('bad request status=401 msg=Authentication failed, check your credentials', done);
     });
+
+    it('should retry - no remaining', function(done) {
```
Could we add some more tests here? I'm thinking:
- Check that a successful request writes the limits to redis
- Check that when there is no data in redis, the request is made
- Check that when there is data in redis and a non-zero remaining, the request is made

maxence-charriere commented:
@tsholmes
Just refactored with your suggestions.

Added a few test for limit and setLimit. 

Also I commented the random condition in limit.
I think the try turnover is already good with the 5 min interval.
Let me know what you think.

tsholmes commented on lib/index.js:
```diff
@@ -48,27 +48,48 @@ var Drip = module.exports = integration('Drip')
+          }
+
+          if (res.status === 429) {
+            err = new Error('too many requests')
```
should be able to just pass `err` out as-is. it's already a 429 error

tsholmes commented on lib/index.js:
```diff
@@ -48,27 +48,48 @@ var Drip = module.exports = integration('Drip')
+          if (res.status === 429) {
+            err = new Error('too many requests')
+            err.status = 429;
+            return self.setLimit(header, function() {
```
i think you want `req.headers` here

tsholmes commented on lib/index.js:
```diff
@@ -113,4 +158,66 @@ Drip.prototype.normalize = function(obj){
+    reset: headers['x-ratelimit-reset'] || Date.now() + (1000 * 60 * 5)
+  };
+
+  redis.set(key, JSON.stringify(limit), function(err, res) {
```
We should probably set an expiration on these keys so they don't live forever. An hour or 2 should be good I think?

https://github.com/segment-integrations/integration-intercom/blob/master/lib/index.js#L280

tsholmes commented on test/index.js:
```diff
@@ -124,4 +141,87 @@ describe('Drip', function(){
+        done();
+      };
+
+      drip.limit(req, null);
```
should be able to just call `drip.limit(done, done);` here. If it errors, it will pass it to `done` and fail the test

tsholmes commented on test/index.js:
```diff
@@ -6,21 +6,38 @@ var assert = require('assert');
+    db.on('error', done);
+  });
+
+  beforeEach(function() {
```
Could we add a `db.flushall` to the beforeEach just to make sure redis is empty when we're starting

tsholmes commented on test/index.js:
```diff
@@ -124,4 +141,87 @@ describe('Drip', function(){
+      var remaining = 0;
+      var headers = {
+        'x-ratelimit-remaining': remaining,
+        'x-ratelimit-reset': Date.now()
```
we should subtract a second here to avoid sporadic failures when it runs too fast

maxence-charriere commented:
Thanks @tsholmes 
Just refactored with your feedbacks.
